// zombie/lib/zombie/document.js
// Generated by CoffeeScript 1.6.2
var HTML, JSDOM, JSDOMSelectors, Path, createDocument;

JSDOM = require("jsdom");

HTML = JSDOM.dom.level3.html;

Path = require("path");

JSDOMSelectors = require(Path.resolve(require.resolve("jsdom"), "../jsdom/selectors/index"));

module.exports = createDocument = function(browser, window, referer) {
  var document, features, jsdomBrowser;

  features = {
    MutationEvents: "2.0",
    ProcessExternalResources: [],
    FetchExternalResources: [],
    QuerySelector: true
  };
  jsdomBrowser = JSDOM.browserAugmentation(HTML, {
    parser: browser.htmlParser
  });
  document = new jsdomBrowser.HTMLDocument({
    referrer: referer
  });
  JSDOMSelectors.applyQuerySelector(document, HTML);
  if (browser.hasFeature("scripts", true)) {
    features.ProcessExternalResources.push("script");
    features.FetchExternalResources.push("script");
  }
  if (browser.hasFeature("css", false)) {
    features.FetchExternalResources.push("css");
  }
  if (browser.hasFeature("iframe", true)) {
    features.FetchExternalResources.push("iframe");
  }
  JSDOM.applyDocumentFeatures(document, features);
  Object.defineProperty(document, "window", {
    value: window
  });
  Object.defineProperty(document, "parentWindow", {
    value: window.parent
  });
  Object.defineProperty(document, "location", {
    get: function() {
      return window.location;
    },
    set: function(url) {
      return window.location = url;
    }
  });
  Object.defineProperty(document, "URL", {
    get: function() {
      return window.location.href;
    }
  });
  return document;
};
